File: actin-common/src/main/java/com/hartwig/actin/clinical/feed/questionnaire/QuestionnaireExtraction.java
package com.hartwig.actin.clinical.feed.questionnaire;

import java.util.List;
import java.util.Map;

import com.google.common.annotations.VisibleForTesting;
import com.google.common.collect.Lists;
import com.hartwig.actin.clinical.datamodel.TumorStage;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.apache.logging.log4j.util.Strings;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public final class QuestionnaireExtraction {

    private static final Logger LOGGER = LogManager.getLogger(QuestionnaireExtraction.class);

    private static final String KEY_VALUE_SEPARATOR = ":";
    private static final String VALUE_LIST_SEPARATOR_1 = ",";
    private static final String VALUE_LIST_SEPARATOR_2 = ";";

    private static final String ACTIN_QUESTIONNAIRE_KEYWORD = "ACTIN Questionnaire";

    private static final int ACTIVE_LINE_OFFSET = 1;
    private static final int SYMPTOMATIC_LINE_OFFSET = 2;

    private QuestionnaireExtraction() {
    }

    public static boolean isActualQuestionnaire(@NotNull QuestionnaireEntry entry) {
        return entry.itemAnswerValueValueString().contains(ACTIN_QUESTIONNAIRE_KEYWORD);
    }

    @Nullable
    public static Questionnaire extract(@Nullable QuestionnaireEntry entry) {
        if (entry == null) {
            return null;
        }

        Map<QuestionnaireKey, String> mapping = QuestionnaireMapping.mapping(entry);

        String significantAberrationLatestECG = value(entry, mapping.get(QuestionnaireKey.SIGNIFICANT_ABERRATION_LATEST_ECG));

        Boolean hasSignificantAberrationLatestECG = significantAberrationLatestECG != null && !significantAberrationLatestECG.isEmpty();
        if (isConfiguredOption(significantAberrationLatestECG)) {
            hasSignificantAberrationLatestECG = toOption(significantAberrationLatestECG);
            if (hasSignificantAberrationLatestECG == null) {
                significantAberrationLatestECG = null;
            } else if (!hasSignificantAberrationLatestECG) {
                significantAberrationLatestECG = Strings.EMPTY;
            }
        }
        return ImmutableQuestionnaire.builder()
                .tumorLocation(value(entry, mapping.get(QuestionnaireKey.PRIMARY_TUMOR_LOCATION)))
                .tumorType(value(entry, mapping.get(QuestionnaireKey.PRIMARY_TUMOR_TYPE)))
                .stage(toStage(value(entry, mapping.get(QuestionnaireKey.STAGE))))
                .treatmentHistoriesCurrentTumor(toList(value(entry, mapping.get(QuestionnaireKey.TREATMENT_HISTORY_CURRENT_TUMOR))))
                .hasMeasurableLesionRecist(toOption(value(entry, mapping.get(QuestionnaireKey.HAS_MEASURABLE_DISEASE_RECIST))))
                .hasBrainLesions(toOption(value(entry, mapping.get(QuestionnaireKey.HAS_BRAIN_LESIONS))))
                .hasActiveBrainLesions(toOption(value(entry, mapping.get(QuestionnaireKey.HAS_BRAIN_LESIONS), ACTIVE_LINE_OFFSET)))
                .hasSymptomaticBrainLesions(toOption(value(entry,
                        mapping.get(QuestionnaireKey.HAS_BRAIN_LESIONS),
                        SYMPTOMATIC_LINE_OFFSET)))
                .hasCnsLesions(toOption(value(entry, mapping.get(QuestionnaireKey.HAS_CNS_LESIONS))))
                .hasActiveCnsLesions(toOption(value(entry, mapping.get(QuestionnaireKey.HAS_CNS_LESIONS), ACTIVE_LINE_OFFSET)))
                .hasSymptomaticCnsLesions(toOption(value(entry, mapping.get(QuestionnaireKey.HAS_CNS_LESIONS), SYMPTOMATIC_LINE_OFFSET)))
                .hasBoneLesions(toOption(value(entry, mapping.get(QuestionnaireKey.HAS_BONE_LESIONS))))
                .hasLiverLesions(toOption(value(entry, mapping.get(QuestionnaireKey.HAS_LIVER_LESIONS))))
                .whoStatus(toWHO(value(entry, mapping.get(QuestionnaireKey.WHO_STATUS))))
                .hasSignificantCurrentInfection(toOption(value(entry, mapping.get(QuestionnaireKey.SIGNIFICANT_CURRENT_INFECTION))))
                .hasSignificantAberrationLatestECG(hasSignificantAberrationLatestECG)
                .significantAberrationLatestECG(significantAberrationLatestECG)
                .build();
    }

    @Nullable
    @VisibleForTesting
    static TumorStage toStage(@Nullable String stage) {
        if (stage == null || stage.isEmpty()) {
            return null;
        }

        if (!QuestionnaireConstants.STAGE_MAPPING.containsKey(stage)) {
            LOGGER.warn("Unrecognized questionnaire tumor stage: '{}'", stage);
            return null;
        }

        return QuestionnaireConstants.STAGE_MAPPING.get(stage);
    }

    @Nullable
    @VisibleForTesting
    static Boolean toOption(@Nullable String option) {
        if (option == null || option.isEmpty()) {
            return null;
        }

        if (!isConfiguredOption(option)) {
            LOGGER.warn("Unrecognized questionnaire option: '{}'", option);
            return null;
        }

        return QuestionnaireConstants.OPTION_MAPPING.get(option);
    }

    private static boolean isConfiguredOption(@Nullable String option) {
        if (option == null) {
            return false;
        }

        return QuestionnaireConstants.OPTION_MAPPING.containsKey(option);
    }

    @Nullable
    @VisibleForTesting
    static Integer toWHO(@Nullable String integer) {
        if (integer == null || integer.isEmpty()) {
            return null;
        }

        int value = Integer.parseInt(integer);
        if (value >= 0 && value <= 4) {
            return value;
        } else {
            LOGGER.warn("Unrecognized WHO status: '{}'", value);
            return null;
        }
    }

    @Nullable
    private static List<String> toList(@Nullable String value) {
        if (value == null) {
            return null;
        }

        String[] split;
        if (value.contains(VALUE_LIST_SEPARATOR_1)) {
            split = value.split(VALUE_LIST_SEPARATOR_1);
        } else {
            split = value.split(VALUE_LIST_SEPARATOR_2);
        }

        return cleanAndTrim(split);
    }

    @NotNull
    private static List<String> cleanAndTrim(@NotNull String[] values) {
        List<String> trimmed = Lists.newArrayList();
        for (String value : values) {
            String trim = value.trim();
            if (!trim.isEmpty()) {
                trimmed.add(trim);
            }
        }
        return trimmed;
    }

    @Nullable
    private static String value(@NotNull QuestionnaireEntry entry, @Nullable String key) {
        return value(entry, key, 0);
    }

    @Nullable
    private static String value(@NotNull QuestionnaireEntry entry, @Nullable String key, int lineOffset) {
        LookupResult result = lookup(entry, key);

        String line = result != null ? result.lines[result.lineIndex + lineOffset] : null;
        return line != null ? line.substring(line.indexOf(KEY_VALUE_SEPARATOR) + 1).trim() : null;
    }

    @Nullable
    private static LookupResult lookup(@NotNull QuestionnaireEntry entry, @Nullable String key) {
        if (key == null) {
            return null;
        }

        String[] lines = read(entry);

        for (int i = 0; i < lines.length; i++) {
            if (lines[i].contains(key)) {
                return new LookupResult(lines, i);
            }
        }

        throw new IllegalStateException("Could not find key " + key + " in questionnaire " + entry);
    }

    @NotNull
    private static String[] read(@NotNull QuestionnaireEntry entry) {
        String[] lines = entry.itemAnswerValueValueString().split("\n");
        String[] cleaned = new String[lines.length];
        for (int i = 0; i < lines.length; i++) {
            String clean = lines[i];
            for (String term : QuestionnaireConstants.TERMS_TO_CLEAN) {
                while (clean.contains(term)) {
                    clean = clean.replace(term, Strings.EMPTY);
                }
            }
            cleaned[i] = clean;
        }
        return cleaned;
    }

    private static class LookupResult {

        @NotNull
        private final String[] lines;
        private final int lineIndex;

        public LookupResult(@NotNull final String[] lines, final int lineIndex) {
            this.lines = lines;
            this.lineIndex = lineIndex;
        }
    }
}

File: clinical/src/main/kotlin/com/hartwig/actin/clinical/feed/emc/questionnaire/QuestionnaireExtraction.kt
package com.hartwig.actin.clinical.feed.emc.questionnaire

import com.google.common.collect.Lists
import org.apache.logging.log4j.LogManager

object QuestionnaireExtraction {
    private val LOGGER = LogManager.getLogger(QuestionnaireExtraction::class.java)
    const val KEY_VALUE_SEPARATOR = ":"
    const val VALUE_LIST_SEPARATOR_1 = ","
    private const val VALUE_LIST_SEPARATOR_2 = ";"
    private const val ACTIN_QUESTIONNAIRE_KEYWORD = "ACTIN Questionnaire"
    private const val ACTIVE_LINE_OFFSET = 1

    fun extract(entry: QuestionnaireEntry?): Pair<Questionnaire?, List<QuestionnaireCurationError>> {
        if (entry == null || !isActualQuestionnaire(entry)) {
            return null to emptyList()
        }
        val mapping = QuestionnaireMapping.mapping(entry)
        val lines = QuestionnaireReader.read(entry.text, QuestionnaireMapping.keyStrings(entry), QuestionnaireMapping.SECTION_HEADERS)
        val brainLesionData = lesionData(entry.subject, lines, mapping[QuestionnaireKey.HAS_BRAIN_LESIONS]!!)
        val cnsLesionData = lesionData(entry.subject, lines, mapping[QuestionnaireKey.HAS_CNS_LESIONS]!!)
        val hasMeasurableDisease =
            QuestionnaireCuration.toOption(entry.subject, value(lines, mapping[QuestionnaireKey.HAS_MEASURABLE_DISEASE]))
        val hasBoneLesions = QuestionnaireCuration.toOption(entry.subject, value(lines, mapping[QuestionnaireKey.HAS_BONE_LESIONS]))
        val hasLiverLesions = QuestionnaireCuration.toOption(entry.subject, value(lines, mapping[QuestionnaireKey.HAS_LIVER_LESIONS]))
        val whoStatus = QuestionnaireCuration.toWHO(entry.subject, value(lines, mapping[QuestionnaireKey.WHO_STATUS]))
        val infectionStatus =
            QuestionnaireCuration.toInfectionStatus(entry.subject, value(lines, mapping[QuestionnaireKey.SIGNIFICANT_CURRENT_INFECTION]))
        val ecg = QuestionnaireCuration.toECG(entry.subject, value(lines, mapping[QuestionnaireKey.SIGNIFICANT_ABERRATION_LATEST_ECG]))
        val stage = QuestionnaireCuration.toStage(entry.subject, value(lines, mapping[QuestionnaireKey.STAGE]))
        return Questionnaire(
            date = entry.authored,
            tumorLocation = value(lines, mapping[QuestionnaireKey.PRIMARY_TUMOR_LOCATION]),
            tumorType = if (QuestionnaireVersion.version(entry) == QuestionnaireVersion.V0_1) "Unknown" else value(
                lines,
                mapping[QuestionnaireKey.PRIMARY_TUMOR_TYPE]
            ),
            biopsyLocation = value(lines, mapping[QuestionnaireKey.BIOPSY_LOCATION]),
            stage = stage.curated,
            treatmentHistoryCurrentTumor = toList(value(lines, mapping[QuestionnaireKey.TREATMENT_HISTORY_CURRENT_TUMOR])),
            otherOncologicalHistory = toList(value(lines, mapping[QuestionnaireKey.OTHER_ONCOLOGICAL_HISTORY])),
            secondaryPrimaries = secondaryPrimaries(lines, mapping[QuestionnaireKey.SECONDARY_PRIMARY]),
            nonOncologicalHistory = toList(value(lines, mapping[QuestionnaireKey.NON_ONCOLOGICAL_HISTORY])),
            ihcTestResults = toList(value(lines, mapping[QuestionnaireKey.IHC_TEST_RESULTS])),
            pdl1TestResults = toList(value(lines, mapping[QuestionnaireKey.PDL1_TEST_RESULTS])),
            hasMeasurableDisease = hasMeasurableDisease.curated,
            hasBrainLesions = brainLesionData.curated?.present(),
            hasActiveBrainLesions = brainLesionData.curated?.active(),
            hasCnsLesions = cnsLesionData.curated?.present(),
            hasActiveCnsLesions = cnsLesionData.curated?.active(),
            hasBoneLesions = hasBoneLesions.curated,
            hasLiverLesions = hasLiverLesions.curated,
            otherLesions = otherLesions(entry, lines, mapping),
            whoStatus = whoStatus.curated,
            unresolvedToxicities = toList(value(lines, mapping[QuestionnaireKey.UNRESOLVED_TOXICITIES])),
            infectionStatus = infectionStatus.curated,
            ecg = ecg.curated,
            complications = toList(value(lines, mapping[QuestionnaireKey.COMPLICATIONS])),
        ) to hasBoneLesions.errors + hasLiverLesions.errors + hasMeasurableDisease.errors + whoStatus.errors + infectionStatus.errors + ecg.errors + stage.errors
    }

    fun isActualQuestionnaire(entry: QuestionnaireEntry): Boolean {
        return entry.text.contains(ACTIN_QUESTIONNAIRE_KEYWORD)
    }

    private fun secondaryPrimaries(lines: Array<String>, secondaryPrimaryKey: String?): List<String>? {
        val extractedValues = if (secondaryPrimaryKey == null) null else values(lines, secondaryPrimaryKey, 1)
        return if (extractedValues == null || extractedValues[0].isEmpty()) {
            null
        } else QuestionnaireCuration.toSecondaryPrimaries(extractedValues[0], extractedValues[1])
    }

    private fun otherLesions(
        entry: QuestionnaireEntry, lines: Array<String>,
        mapping: Map<QuestionnaireKey, String?>
    ): List<String>? {
        val version: QuestionnaireVersion = QuestionnaireVersion.version(entry)
        return if (version == QuestionnaireVersion.V0_1) {
            //In v0.1, the format for primary tumor location is "$location ($otherLesions)"
            val input = value(lines, mapping[QuestionnaireKey.PRIMARY_TUMOR_LOCATION])
            if (input != null && input.contains("(") && input.contains(")")) {
                val start = input.indexOf("(")
                val end = input.indexOf(")")
                toList(input.substring(start + 1, end))
            } else {
                null
            }
        } else {
            toList(value(lines, mapping[QuestionnaireKey.OTHER_LESIONS]))
        }
    }

    private fun lesionData(subject: String, lines: Array<String>, keyString: String): ValidatedQuestionnaireCuration<LesionData> {
        val extractedValues = values(lines, keyString, ACTIVE_LINE_OFFSET)
        return if (extractedValues == null) ValidatedQuestionnaireCuration(LesionData(null, null)) else LesionData.fromString(
            subject,
            extractedValues[0],
            extractedValues[1]
        )
    }

    private fun toList(value: String?): List<String>? {
        if (value == null) {
            return null
        }
        val reformatted = value.replace(VALUE_LIST_SEPARATOR_2.toRegex(), VALUE_LIST_SEPARATOR_1)
        val split = reformatted.split(VALUE_LIST_SEPARATOR_1.toRegex()).dropLastWhile { it.isEmpty() }
            .toTypedArray()
        return cleanAndTrim(split)
    }

    private fun cleanAndTrim(values: Array<String>): List<String> {
        val trimmed: MutableList<String> = Lists.newArrayList()
        for (value in values) {
            val trim = value.trim { it <= ' ' }
            if (trim.isNotEmpty()) {
                trimmed.add(trim)
            }
        }
        return trimmed
    }

    private fun value(lines: Array<String>, key: String?, isOptional: Boolean = false): String? {
        val lineIndex = lookup(lines, key, isOptional) ?: return null
        val extracted = extractValue(lines[lineIndex])
        return extracted.ifEmpty { null }
    }

    private fun values(lines: Array<String>, key: String?, lineOffset: Int): List<String>? {
        val lineIndex = lookup(lines, key, false)
        if (lineIndex != null) {
            val extractedValues = lines.slice(lineIndex..lineIndex + lineOffset)
                .map { line: String -> extractValue(line) }
            if (extractedValues.size < lineOffset + 1) {
                throw RuntimeException(String.format("Failed to extract %d lines for key '%s'", lineOffset + 1, key))
            }
            return extractedValues
        }
        return null
    }

    private fun extractValue(line: String): String {
        return line.substring(line.indexOf(KEY_VALUE_SEPARATOR) + 1).trim { it <= ' ' }
    }

    private fun lookup(lines: Array<String>, key: String?, isOptional: Boolean): Int? {
        if (key == null) {
            return null
        }
        for (i in lines.indices) {
            if (lines[i].contains(key)) {
                return i
            }
        }
        if (isOptional) {
            LOGGER.debug(
                "Key '{}' not present but skipped since it is configured as optional in questionnaire '{}'", key,
                lines.joinToString("\n")
            )
            return null
        }
        throw IllegalStateException("Could not find key '$key' in questionnaire: " + lines.joinToString("\n"))
    }
}